### PR TITLE
Fix implementation of index_put | fix(torchlib)

### PR DESCRIPTION
**Pull request by justinchuby**
Continuation of #1277 by @xadupre 

The onnx implementation of index_put is different in torch script exporter
(https://github.com/pytorch/pytorch/blob/main/torch/onnx/symbolic_opset11.py#L212). The PR replaces the current implementation failing on one corner case by the one from torch script.

### CODE REVIEW

**Comment by justinchuby** [comment]
I get

### Summary

ONNX Runtime raises `[ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Non-zero status code returned while running If node. Name:'_inlfunc_aten_index_put_n8' Status Message: Non-zero status code returned while running ScatterND node. Name:'_inlfunc_aten_index_put_n2' Status Message: CPU execution provider: MLFloat16 data type is not supported with ScatterND opset 16 when reduction is 'add'.` when executing test `ops_test.TestOutputConsistencyFullGraphCPU.test_output_match_opinfo__index_put_cpu_float16` in ONNX Script `TorchLib`.

To recreate this report, use

```bash
CREATE_REPRODUCTION_REPORT=1 python -m pytest onnxscript/tests/function_libs/torch_lib/ops_test.py -k test_output_match_opinfo__index_put_cpu_float16
```

### To reproduce

```python
import google.protobuf.text_format
import numpy as np
from numpy import array, float16, float32, float64, int32, int64
import onnx
import onnxruntime as ort

# Run n times
N = 1

onnx_model_text = """
ir_version: 8
producer_name: "pytorch"
producer_version: "2.3.0"
graph {
  node {
    input: "input_1_0"
    output: "3"
    name: "SequenceConstruct_0"
    op_type: "SequenceConstruct"
    doc_string: ""
  }
  node {
    input: "input_0"
    input: "3"
    input: "input_2"
    output: "_val_4"
    name: "aten_index_put_1"
    op_type: "aten_index_put"
    attribute {
      name: "accumulate"
      i: 1
      type: INT
    }
    doc_string: ""
    domain: "pkg.onnxscript.torch_lib"
  }
  name: "main_graph"
  input {
    name: "input_0"
    type {
      tensor_type {
        elem_type: 10
        shape {
          dim {
            dim_value: 5
          }
          dim {
            dim_value: 5
          }
        }
      }
    }
  }
  input {
    name: "input_1_0"
    type {
      tensor_type {
        elem_type: 7
        shape {
          dim {
            dim_value: 2
          }
        }
      }
    }
  }
  input {
    name: "input_2"
    type {
      tensor_type {
        elem_type: 10
        shape {
          dim {
            dim_value: 2
          }
          dim {
            dim_value: 5
          }
        }
      }
    }
  }
  output {
    name: "_val_4"
    type {
      tensor_type {
        elem_type: 10
        shape {
          dim {
            dim_value: 5
          }
          dim {
            dim_value: 5
          }
        }
      }
    }
  }
  value_info {
    name: "input_0"
    type {
      tensor_type {
        elem_type: 10
        shape {
          dim {
            dim_value: 5
          }
          dim {
            dim_value: 5
          }
        }
      }
    }
  }
  value_info {
    name: "input_1_0"
    type {
      tensor_type {
        elem_type: 7
        shape {
          dim {
            dim_value: 2
          }
        }
      }
    }
  }
  value_info {
    name: "input_2"
    type {
      tensor_type {
        elem_type: 10
        shape {
          dim {
            dim_value: 2
          }
          dim {
            dim_value: 5
          }
        }
      }
    }
  }
  value_info {
    name: "_val_4"
    type {
      tensor_type {
        elem_type: 10
        shape {
          dim {
            dim_value: 5
          }
          dim {
            dim_value: 5
          }
        }
      }
    }
  }
  value_info {
    name: "3"
    type {
      sequence_type {
        elem_type {
          tensor_type {
            elem_type: 7
            shape {
              dim {
                dim_value: 2
              }
            }
          }
        }
      }
    }
  }
}
opset_import {
  domain: "pkg.onnxscript.torch_lib"
  version: 1
}
opset_import {
  domain: ""
  version: 18
}
opset_import {
  domain: "pkg.onnxscript.torch_lib.common"
  version: 1
}
functions {
  name: "aten_index_put"
  input: "self"
  input: "indices"
  input: "values"
  output: "result_2"
  node {
    output: "int64_0"
    name: "n0"
    op_type: "Constant"
    attribute {
      name: "value"
      t {
        data_type: 7
        int64_data: 0
        name: "int64_0"
      }
      type: TENSOR
    }
    domain: ""
  }
  node {
    input: "indices"
    input: "int64_0"
    output: "index"
    name: "n1"
    op_type: "SequenceAt"
    domain: ""
  }
  node {
    output: "int64_m1_1d"
    name: "n2"
    op_type: "Constant"
    attribute {
      name: "value"
      t {
        dims: 1
        data_type: 7
        int64_data: -1
        name: "int64_m1_1d"
      }
      type: TENSOR
    }
    domain: ""
  }
  node {
    input: "index"
    input: "int64_m1_1d"
    output: "new_index"
    name: "n3"
    op_type: "Unsqueeze"
    domain: ""
  }
  node {
    input: "self"
    output: "shape_self"
    name: "n4"
    op_type: "Shape"
    domain: ""
  }
  node {
    output: "accumulate"
    name: "n5"
    op_type: "Constant"
    attribute {
      name: "value_int"
      type: INT
      ref_attr_name: "accumulate"
    }
    domain: ""
  }
  node {
    input: "accumulate"
    output: "accumulate_as_bool"
    name: "n6"
    op_type: "Cast"
    attribute {
      name: "to"
      i: 9
      type: INT
    }
    domain: ""
  }
  node {
    input: "accumulate_as_bool"
    output: "cond"
    name: "n7"
    op_type: "Cast"
    attribute {
      name: "to"
      i: 9
      type: INT
    }
    domain: ""
  }
  node {
    input: "cond"
    output: "result_2"
    name: "n8"
    op_type: "If"
    attribute {
      name: "then_branch"
      g {
        node {
          input: "shape_self"
          output: "tmp"
          name: "n0"
          op_type: "ConstantOfShape"
          domain: ""
        }
        node {
          input: "tmp"
          input: "values"
          output: "zeros"
          name: "n1"
          op_type: "CastLike"
          domain: ""
        }
        node {
          input: "zeros"
          input: "new_index"
          input: "values"
          output: "result"
          name: "n2"
          op_type: "ScatterND"
          attribute {
            name: "reduction"
            s: "add"
            type: STRING
          }
          domain: ""
        }
        node {
          input: "result"
          input: "self"
          output: "result_0"
          name: "n3"
          op_type: "Add"
          domain: ""
        }
        name: "thenGraph_18"
        output {
          name: "result_0"
          type {
          }
        }
      }
      type: GRAPH
    }
    attribute {
      name: "else_branch"
      g {
        node {
          input: "self"
          input: "new_index"
          input: "values"
          output: "result_1"
          name: "n0"
          op_type: "ScatterND"
          domain: ""
        }
        name: "elseGraph_18"
        output {
          name: "result_1"
          type {
          }
        }
      }
      type: GRAPH
    }
    domain: ""
  }
  doc_string: "index_put(Tensor self, Tensor?[] indices, Tensor values, bool accumulate=False) -> Tensor\n    See implementation of `torch.onnx.symbolic_opset11.index_put\n    <https://github.com/pytorch/pytorch/blob/main/torch/onnx/symbolic_opset11.py#L212>`_.\n    "
  opset_import {
    domain: ""
    version: 18
  }
  domain: "pkg.onnxscript.torch_lib"
  attribute_proto {
    name: "accumulate"
    i: 0
    type: INT
  }
}
functions {
  name: "Rank"
  input: "input"
  output: "return_val"
  node {
    input: "input"
    output: "tmp"
    name: "n0"
    op_type: "Shape"
    domain: ""
  }
  node {
    input: "tmp"
    output: "return_val"
    name: "n1"
    op_type: "Size"
    domain: ""
  }
  doc_string: "Take the rank of the input tensor."
  opset_import {
    domain: ""
    version: 18
  }
  domain: "pkg.onnxscript.torch_lib.common"
}
functions {
  name: "IsScalar"
  input: "input"
  output: "return_val"
  node {
    input: "input"
    output: "tmp"
    name: "n0"
    op_type: "Shape"
    domain: ""
  }
  node {
    input: "tmp"
    output: "tmp_0"
    name: "n1"
    op_type: "Size"
    domain: ""
  }
  node {
    output: "tmp_1"
    name: "n2"
    op_type: "Constant"
    attribute {
      name: "value_int"
      i: 0
      type: INT
    }
    domain: ""
  }
  node {
    input: "tmp_0"
    input: "tmp_1"
    output: "return_val"
    name: "n3"
    op_type: "Equal"
    domain: ""
  }
  doc_string: "Return whether the input has rank 0, or is a scalar."
  opset_import {
    domain: ""
    version: 18
  }
  domain: "pkg.onnxscript.torch_lib.common"
}

"""

ort_inputs = {'input_0': array([[ 4.93   ,  7.656  ,  3.402  ,  0.00879, -0.02637],
       [ 3.418  , -4.035  ,  0.923  ,  6.777  ,  7.215  ],
       [ 4.184  , -2.83   , -5.477  , -2.594  ,  4.88   ],
       [-7.586  , -7.234  ,  8.414  , -0.255  , -6.637  ],
       [ 7.58   , -1.837  ,  2.373  , -5.     ,  4.05   ]], dtype=float16), 'input_1_0': array([1, 3]), 'input_2': array([[-6.406,  6.363, -2.68 ,  6.574, -6.04 ],
       [-1.283,  0.457, -2.047,  8.09 , -7.1  ]], dtype=float16)}

# Set up the inference session
session_options = ort.SessionOptions()
session_options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_DISABLE_ALL
onnx_model = onnx.ModelProto()
google.protobuf.text_format.Parse(onnx_model_text, onnx_model)

# Uncomment this line to save the model to a file for examination
# onnx.save_model(onnx_model, "test_output_match_opinfo__index_put_cpu_float16.onnx")

onnx.checker.check_model(onnx_model)
session = ort.InferenceSession(onnx_model.SerializeToString(), session_options, providers=("CPUExecutionProvider",))

# Run the model
for _ in range(N):
    ort_outputs = session.run(None, ort_inputs)

```

### Full error stack

```
[ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Non-zero status code returned while running If node. Name:'_inlfunc_aten_index_put_n8' Status Message: Non-zero status code returned while running ScatterND node. Name:'_inlfunc_aten_index_put_n2' Status Message: CPU execution provider: MLFloat16 data type is not supported with ScatterND opset 16 when reduction is 'add'.
  File "/home/justinchu/dev/onnx-script/onnxscript/tests/function_libs/torch_lib/ops_test_common.py", line 548, in _capture_graph_and_evaluate_torch_script_evaluator
    return _safe_ort_session_run(onnx_model.SerializeToString(), ort_inputs)
  File "/home/justinchu/dev/onnx-script/onnxscript/tests/function_libs/torch_lib/ops_test_common.py", line 363, in _safe_ort_session_run
    raise return_dict["error"]

```

### The ONNX model text for visualization

```
<
   ir_version: 8,
   opset_import: ["pkg.onnxscript.torch_lib" : 1, "" : 18, "pkg.onnxscript.torch_lib.common" : 1],
   producer_name: "pytorch",
   producer_version: "2.3.0"
>
main_graph (float16[5,5] input_0, int64[2] input_1_0, float16[2,5] input_2) => (float16[5,5] _val_4) 
   <float16[5,5] input_0, int64[2] input_1_0, float16[2,5] input_2, float16[5,5] _val_4, seq(int64[2]) 3>
{
   3 = SequenceConstruct (input_1_0)
   _val_4 = pkg.onnxscript.torch_lib.aten_index_put <accumulate: int = 1> (input_0, 3, input_2)
}
<
  domain: "pkg.onnxscript.torch_lib",
  opset_import: ["" : 18]
>
aten_index_put (self, indices, values) => (result_2)
{
   int64_0 = Constant <value: tensor = int64 int64_0 {0}> ()
   index = SequenceAt (indices, int64_0)
   int64_m1_1d = Constant <value: tensor = int64[1] int64_m1_1d {-1}> ()
   new_index = Unsqueeze (index, int64_m1_1d)
   shape_self = Shape (self)
   accumulate = Constant <value_int: int = @accumulate> ()
   accumulate_as_bool = Cast <to: int = 9> (accumulate)
   cond = Cast <to: int = 9> (accumulate_as_bool)
   result_2 = If (cond) <then_branch: graph = thenGraph_18 () => ( result_0) {
      tmp = ConstantOfShape (shape_self)
      zeros = CastLike (tmp, values)
      result = ScatterND <reduction: string = "add"> (zeros, new_index, values)
      result_0 = Add (result, self)
   }, else_branch: graph = elseGraph_18 () => ( result_1) {
      result_1 = ScatterND (self, new_index, values)
   }>
}
<
  domain: "pkg.onnxscript.torch_lib.common",
  opset_import: ["" : 18]
>
Rank (input) => (return_val)
{
   tmp = Shape (input)
   return_val = Size (tmp)
}
<
  domain: "pkg.onnxscript.torch_lib.common",
  opset_import: ["" : 18]
>
IsScalar (input) => (return_val)
{
   tmp = Shape (input)
   tmp_0 = Size (tmp)
   tmp_1 = Constant <value_int: int = 0> ()
   return_val = Equal (tmp_0, tmp_1)
}
```

### Environment

```
OS: Linux-6.5.0-1016-azure-x86_64-with-glibc2.35
Python version: 3.10.9 (main, Jan 11 2023, 15:21:40) [GCC 11.2.0]
onnx==1.16.0.dev20240205
onnxruntime==1.17.0
numpy==1.25.1
torch==2.3.0.dev20240305+cpu
```

**Comment by justinchuby** [comment]
@xadupre with the added test, the PyTorch operator itself raises "shape mismatch: value tensor of shape [2, 3, 4] cannot be broadcast to indexing result of shape [4]". Was there an adjustment that can be made to correct the test?

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 1]
```diff
@@ -4065,27 +4065,24 @@ def aten_index_put(
...
+    shape_self = op.Shape(self)
 
     if op.Cast(accumulate, to=BOOL.dtype):
-        # put values into zeros array first, then add to input
-        zeros = op.Expand(op.Constant(value_float=0.0), op.Shape(self))
-        zeros = op.CastLike(zeros, values)
-        result = op.ScatterElements(zeros, new_ind_t, values)
-        # FIXME: type promotion
-        result = op.CastLike(result, self)
+        zeros = op.CastLike(op.ConstantOfShape(shape_self), values)
```
Comment (xadupre) [inline onnxscript/function_libs/torch_lib/ops/core.py:4080]: Maybe we can remove CastLike and operator Add and keep almost the same code for both path of the test. reduction was added in opset 13.

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:4073]: Do you mean we always do `result = op.ScatterND(self, new_index, values)`?

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 2]
```diff
@@ -4058,27 +4058,24 @@ def aten_index_put(
...
+        result = op.ScatterND(zeros, new_index, values, reduction="add")
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4074]:
```suggestion
        result = op.ScatterND(result, new_index, values, reduction="add")
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 3]
```diff
@@ -4058,27 +4058,23 @@ def aten_index_put(
...
 
     if op.Cast(accumulate, to=BOOL.dtype):
-        # put values into zeros array first, then add to input
-        zeros = op.Expand(op.Constant(value_float=0.0), op.Shape(self))
-        zeros = op.CastLike(zeros, values)
-        result = op.ScatterElements(zeros, new_ind_t, values)
-        # FIXME: type promotion
-        result = op.CastLike(result, self)
+        result = op.ScatterND(result, new_index, values, reduction="add")
         result = op.Add(result, self)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4074]:
```suggestion
        result = op.ScatterND(result, new_index, values, reduction="add")
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 4]
```diff
@@ -4058,27 +4058,22 @@ def aten_index_put(
...
 
     if op.Cast(accumulate, to=BOOL.dtype):
-        # put values into zeros array first, then add to input
-        zeros = op.Expand(op.Constant(value_float=0.0), op.Shape(self))
-        zeros = op.CastLike(zeros, values)
-        result = op.ScatterElements(zeros, new_ind_t, values)
-        # FIXME: type promotion
-        result = op.CastLike(result, self)
-        result = op.Add(result, self)
+        result = op.ScatterND(result, new_index, values, reduction="add")
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4073]:
```suggestion
        result = op.ScatterND(self, new_index, values, reduction="add")
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 5]
```diff
@@ -4058,27 +4058,22 @@ def aten_index_put(
...
-    new_ind = op.Expand(index, shape)
-    new_ind_t = op.Transpose(new_ind)
+    See implementation of `torch.onnx.symbolic_opset11.index_put
+    <https://github.com/pytorch/pytorch/blob/main/torch/onnx/symbolic_opset11.py#L212>`_.
+    """
+
+    # TODO(justinchuby): Handle when indicies has more than one element
+    index = op.SequenceAt(indices, 0)
+    new_index = op.Unsqueeze(index, [-1])
+    shape_self = op.Shape(self)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4070]:
```suggestion
```

**Review by xadupre — APPROVED** [review]

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py]: Done

**Pull request merged**